packages/tailwindcss/src/at-import.ts

```diff
@@ -0,0 +1,144 @@
+  base: string,
+  loadStylesheet: LoadStylesheet,
+  recurseCount = 0,
+) {
+  let promises: Promise<void>[] = []
+
+  walk(ast, (node, { replaceWith }) => {
+    if (
+      node.kind === 'rule' &&
+      node.selector[0] === '@' &&
+      node.selector.toLowerCase().startsWith('@import ')
+    ) {
+      try {
+        let { uri, layer, media, supports } = parseImportParams(
+          ValueParser.parse(node.selector.slice(8)),
+        )
+
+        // Skip importing data or remote URIs
+        if (uri.startsWith('data:')) return
+        if (uri.startsWith('http://') || uri.startsWith('https://')) return
+
+        let contextNode = context({}, [])
+
+        promises.push(
+          (async () => {
+            // Since we do not have fully resolved paths in core, we can't reliably detect circular
+            // imports. Instead, we try to limit the recursion depth to a number that is too large
+            // to be reached in practice.
+            if (recurseCount > 100) {
+              throw new Error(
+                `Exceeded maximum recursion depth while resolving \`${uri}\` in \`${base}\`)`,
+              )
+            }
+
+            const loaded = await loadStylesheet(uri, base)
+            let ast = CSS.parse(loaded.content)
+            await substituteAtImports(ast, loaded.base, loadStylesheet, recurseCount + 1)
+
+            contextNode.nodes = buildImportNodes(ast, layer, media, supports)
+            contextNode.context.base = loaded.base
+          })(),
+        )
+
+        replaceWith(contextNode)
+        // The resolved Stylesheets already have their transitive @imports
+        // resolved, so we can skip walking them.
+        return WalkAction.Skip
+      } catch (e: any) {
+        // When an error occurs while parsing the `@import` statement, we skip
+        // the import.
+      }
+    }
+  })
+
+  await Promise.all(promises)
+}
+
+// Modified and inlined version of `parse-statements` from
+// `postcss-import` <https://github.com/postcss/postcss-import>
+// Copyright (c) 2014 Maxime Thirouin, Jason Campbell & Kevin Mårtensson
+// Released under the MIT License.
+function parseImportParams(params: ValueParser.ValueAstNode[]) {
+  let uri
+  let layer: string | null = null
+  let media: string | null = null
+  let supports: string | null = null
+
+  for (let i = 0; i < params.length; i++) {
+    const node = params[i]
+
+    if (node.kind === 'separator') continue
+
+    if (node.kind === 'word' && !uri) {
+      if (!node.value) throw new Error(`Unable to find uri`)
+      if (node.value[0] !== '"' && node.value[0] !== "'") throw new Error('Unable to find uri')
+
+      uri = node.value.slice(1, -1)
+      continue
+    }
+
+    if (node.kind === 'function' && /^url$/i.test(node.value)) {
+      throw new Error('url functions are not supported')
+    }
+
+    if (!uri) throw new Error('Unable to find uri')
+
+    if ((node.kind === 'word' || node.kind === 'function') && /^layer$/i.test(node.value)) {
```

Comment:
Should we precompile these regular expressions? Not a super hot path I don't think but feel like we're consistent about that elsewhere.

Reply:
Noticed these are also full string comparisons, I guess only using a regex to make the comparison case-insensitive? Wonder how `.toLowercase() === ...` compares, if perf is the ~same then we can just not use a regex and don't need to think about precompiling it.

Reply:
toLowerCase makes more sense and is also faster 👍